### PR TITLE
Fatjar

### PR DESCRIPTION
**Pull request by WindLeeWT**
#### Generate fat jar with the following rules:
1. Create a simple manifest.mf
2. For each individual jar:
  * META-INF
    * MANIFEST.MF: exclude
    * INDEX.LIST: exclude
    * *.SF/*.DSA/*.RSA/SIG-*: exclude
    * services: include and warn if duplicate
    * others: include and warn if duplicate
  * LICENSE/README/NOTICE related files: exclude
  * others: include and warn if duplicate

### CODE REVIEW

File: src/blade/scons_helper.py [hunk 1]
```diff
@@ -532,6 +532,22 @@ def generate_one_jar(target, source, env):
                              _one_jar_boot_path)
 
 
+def _signature_file(name):
```
Comment (chen3feng) [inline src/blade/scons_helper.py]: _is_

File: src/blade/scons_helper.py [hunk 2]
```diff
@@ -540,32 +556,45 @@ def _generate_fat_jar(target, deps_jar, env):
...
-                # TODO(wentingli): Create manifest from dependency jars
-                if first_jar:
-                    main_class = env.Dictionary().get('JAVAMAINCLASS')
-                    if main_class:
-                        target_fat_jar.writestr(manifest,
-                            'Manifest-Version: 1.0\nMain-Class: %s\n\n' % main_class)
-                    else:
-                        target_fat_jar.writestr(name, jar.read(name))
-                    first_jar = False
+            exclude = False
```
Comment (chen3feng) [inline src/blade/scons_helper.py]: 写个单独的函数
_is_fat_jar_excluded()

File: src/blade/scons_helper.py [hunk 3]
```diff
@@ -540,32 +556,45 @@ def _generate_fat_jar(target, deps_jar, env):
...
+                    zip_path_conflicts += 1
+                    global option_verbose
+                    if option_verbose and jar.read(name):
+                        console.warning('%s: duplicate path %s found in {%s, %s}' % (
+                            os.path.basename(target), name,
+                            zip_path_dict[name], os.path.basename(dep_jar)))
 
         jar.close()
 
+    console.warning('%s: Found %d conflicts when packaging' % (
```
Comment (chen3feng) [inline src/blade/scons_helper.py]: ", --verbose to see details"

File: src/blade/scons_helper.py [hunk 4]
```diff
@@ -540,32 +556,45 @@ def _generate_fat_jar(target, deps_jar, env):
...
             else:
-                if name not in zip_path_set:
+                if name not in zip_path_dict:
                     target_fat_jar.writestr(name, jar.read(name))
-                    zip_path_set.add(name)
+                    zip_path_dict[name] = os.path.basename(dep_jar)
+                else:
+                    zip_path_conflicts += 1
+                    global option_verbose
+                    if option_verbose and jar.read(name):
```
Comment (chen3feng) [inline src/blade/scons_helper.py]: read是为了判断非空文件？感觉没必要。

File: src/blade/scons_helper.py [hunk 5]
```diff
@@ -540,32 +556,45 @@ def _generate_fat_jar(target, deps_jar, env):
...
+                        console.warning('%s: duplicate path %s found in {%s, %s}' % (
```
Comment (chen3feng) [inline src/blade/scons_helper.py]: 这里还可以优化一下，内容完全相同的可以不打，如果能减少很多的话。

File: src/blade/scons_helper.py (hunk 4 above)
Comment (WindLeeWT) [inline src/blade/scons_helper.py]: 这里为了判断目录entry，改用结尾是'/'来判断

File: src/blade/scons_helper.py [hunk 6]
```diff
@@ -532,6 +532,32 @@ def generate_one_jar(target, source, env):
...
+    return False
+
+
+_FATJAR_EXCLUSIONS = frozenset(['LICENSE', 'README', 'NOTICE',
+                                'META-INF/LICENSE', 'META-INF/README',
+                                'META-INF/NOTICE', 'META-INF/INDEX.LIST'])
+
+
+def _is_fat_jar_excluded(name):
+    manifest = os.path.join('META-INF', 'MANIFEST.MF')
```
Comment (chen3feng) [inline src/blade/scons_helper.py]: 这个出现两次了，可以定义成常量

**Pull request merged**